spec.md

```diff
@@ -499,6 +499,54 @@ the module.  The `c`{.firrtl} sub-field contained in the `b`{.firrtl} sub-field
 flows into the module, and the `d`{.firrtl} sub-field contained in the
 `b`{.firrtl} sub-field flows out of the module.
 
+## Type Modifiers
+
+### Constant Type
+
+A constant type is a type whose value is guaranteed to be a compile-time 
+constant.  Constant types may be used in ports, wire, nodes, and generally 
+anywhere a non-constant type is usable.  Operations on constant type are well 
+defined.  As a general rule (with any exception listed in the definition for 
+such operations as have exceptions), an operation whose arguments are constant
+produces a constant.  An operations with some non-constant arguments produce a
+non-constant.  Constants can be used in any context with a source flow which 
+allows a non-constant.  Constants may be used as the target of a connect so long 
+as the source of the connect is itself constant.  These rules ensure all 
+constants are derived from literals or from constant-typed input ports of the 
+top-level module.
```

Comment:
Shouldn't the toplevel be forbidden from featuring constant type input ports? Since then we would not be able to tell the value at compile time?

Reply:
This is intentionally allowed. 

The language here is intending that a value is known to be a constant value to the compiler, not that the compiler knows the value.

This is intended to support a use case of two separately compiled circuits that are then linked at FIRRTL or Verilog levels. Without further FIRRTL spec modifications, this has to be represented as a boundary where an extmodule has a const port and a top-level module also has a const port.

Reply:
Then I don't think that it is correct to call these `compile time` constants. They are more `runtime constants`, i.e. the value does not change while the circuit is executing, but the compiler does not actually know the value.

Reply:
To make the comparison to C++ again: You can have a `extern const int a;` (which is resolved at link time, just like your example) but not a `extern constexpr int a;` (which would be a compile time constant expression).

Reply:
Yeah I think I agree with @ekiwi , they are more runtime constants. Maybe the wording could be clarified?

Reply:
Maybe it's clearer to use the term "literal", rather than "const" or "constant" ?

Reply:
Wording hopefully clarified as to what is meant by "constant".